src/lua/init.lua

```diff
@@ -86,77 +114,85 @@ local function load_lua(file)
     return loadfile(file)
 end
 
-local function search_cwd_lib(name)
-    local path = "./?."..soext
-    return package.searchpath(name, path)
-end
+local function merge_strings(paths, templates)
+    local searchpaths = {}
+    paths = type(paths) == 'table' and paths or { paths }
+    templates = templates or {}
 
-local function search_cwd_lua(name)
-    local path = "./?.lua;./?/init.lua"
-    return package.searchpath(name, path)
-end
+    if #templates == 0 then
+        return paths
+    end
 
-local function traverse_rocks(name, pathes_search)
-    local cwd = fio.cwd()
-    local index = string.len(cwd) + 1
-    local strerr = ""
-    while index ~= nil do
-        cwd = string.sub(cwd, 1, index - 1)
-        for i, path in ipairs(pathes_search) do
-            local file, err = package.searchpath(name, cwd .. path)
-            if err == nil then
-                return file
-            end
-            strerr = strerr .. err
+    if #paths == 0 then
+        return templates
+    end
+
+    for _, path in ipairs(paths) do
+        for _, template in pairs(templates) do
+            table.insert(searchpaths, fio.pathjoin(path, template))
         end
-        index = string.find(cwd, "/[^/]*$")
     end
-    return nil, strerr
+
+    return searchpaths
 end
 
-local function search_rocks_lua(name)
-    local pathes_search = {
-        "/.rocks/share/tarantool/?.lua;",
-        "/.rocks/share/tarantool/?/init.lua;",
-    }
-    return traverse_rocks(name, pathes_search)
+local LIB_TEMPLATES = { '?.'..soext }
+local LUA_TEMPLATES = { '?.lua', '?/init.lua' }
+local ROCKS_LIB_TEMPLATES = { ROCKS_LIB_PATH .. '/?.'..soext }
+local ROCKS_LUA_TEMPLATES = { ROCKS_LUA_PATH .. '/?.lua', ROCKS_LUA_PATH .. '/?/init.lua' }
+
+local function search_lib(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LIB_TEMPLATES), ';'))
 end
 
-local function search_rocks_lib(name)
-    local pathes_search = {
-        "/.rocks/lib/tarantool/?."..soext
-    }
-    return traverse_rocks(name, pathes_search)
+local function search_lua(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LUA_TEMPLATES), ';'))
+end
+
+local function split_path(path)
+    path = path:sub(-1) == '/' and path:sub(1, -2) or path
+    local index = string.len(path) + 1
+    local paths = {}
+
+    while index ~= nil do
+        path = string.sub(path, 1, index - 1)
+        table.insert(paths, path .. '/')
+        index = string.find(path, "/[^/]*$")
+    end
+
+    return paths
 end
```

Comment:
How about calling `fio.dirname` in a loop instead of using regex hacks?

Reply:
Also maybe it should be renamed to `traverse_path`?

Reply:
I agree. I switched to `fio.dirname` and `traverse_path` seems more appropriate name here.